servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/WhenFinallyHttpOperator.java

```diff
@@ -0,0 +1,521 @@
+/*
+ * Copyright © 2019-2025 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.PublisherSource.Subscriber;
+import io.servicetalk.concurrent.PublisherSource.Subscription;
+import io.servicetalk.concurrent.SingleSource;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.SingleOperator;
+import io.servicetalk.concurrent.api.TerminalSignalConsumer;
+import io.servicetalk.concurrent.internal.CancelImmediatelySubscriber;
+import io.servicetalk.http.api.StreamingHttpResponse;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.utils.internal.ThrowableUtils.addSuppressed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+class WhenFinallyHttpOperator implements SingleOperator<StreamingHttpResponse, StreamingHttpResponse> {
```

Comment:
this class is a copy of `BeforeFinallyHttpOperator` that has been modified to trigger events in an order dependent on the `afterFinallyBehavior` parameter. Look for usages of `afterFinallyBehavior` and check the if/else blocks: the new behavior is in the if blocks while old behavior is in the else blocks.